### PR TITLE
Modernize RSpec configuration

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
This PR enables some best-practice RSpec options (for reference, see the project initializer [boilerplate](https://github.com/rspec/rspec-core/blob/main/lib/rspec/core/project_initializer/spec/spec_helper.rb) in `rspec-core`). Specifically, it:

- Enables and fixes warnings (first commit). These involved a circular `require` and re-defining methods without undefining them first. (The latter can seem a bit silly, but I've found to be very useful in understanding code on occasion.)
- Enables and fixes `verify_partial_double` (second commit). This checks that the underlying object actually defines the doubled methods. This uncovered some cruft, though I had to use the escape hatch of `without_partial_double_verification` around ENV code that I couldn't make immediate sense of.
- Enables and fixes `disable_monkey_patching` (remaining commits). This [setting](https://rspec.info/features/3-12/rspec-core/configuration/zero-monkey-patching-mode/) removes RSpec methods from the top-level scope.

### CODE REVIEW

File: Library/Homebrew/extend/os/mac/dev-cmd/bottle.rb
Comment (reitermarkus) [inline Library/Homebrew/extend/os/mac/dev-cmd/bottle.rb:16]:
```suggestion
    undef gnu_tar
    
```

File: Library/Homebrew/extend/os/mac/dev-cmd/bottle.rb
Comment (reitermarkus) [inline Library/Homebrew/extend/os/mac/dev-cmd/bottle.rb:6]:
```suggestion
    undef tar_args

```

File: Library/Homebrew/test/spec_helper.rb
Comment (reitermarkus) [inline Library/Homebrew/test/spec_helper.rb:203]: This can be defined outside the block. Probable best to put it in a `Test::Helper` module.

File: Library/Homebrew/test/spec_helper.rb
Comment (reitermarkus) [inline Library/Homebrew/test/spec_helper.rb:70]: Can we keep the “expose DSL” setting while disabling all other monkey patching? I don't really see the top-level `describe` being a problem since we don't ever call any top-level methods.

Adding 

```ruby
    mocks.syntax = :expect
    mocks.patch_marshal_to_support_partial_doubles = false
```

and 

```ruby
    expectations.syntax = :expect
```

below should do it, according to [https://rubydoc.info/github/rspec/rspec-core/RSpec%2FCore%2FConfiguration:disable_monkey_patching!](https://rubydoc.info/github/rspec/rspec-core/RSpec%2FCore%2FConfiguration:disable_monkey_patching!).

**Review by MikeMcQuaid — APPROVED** [review]
Looks great, thanks again @dduugg!

File: Library/Homebrew/test/spec_helper.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/test/spec_helper.rb:70]: > I don't really see the top-level `describe` being a problem since we don't ever call any top-level methods.

I like the explicitness of avoiding it.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, thanks @dduugg! Remaining feedback can be addressed in follow-up PR, let's get this merged before it conflicts.

**Comment by MikeMcQuaid** [comment]
@dduugg Looks like some legit test failures post-rebase, sorry!

**Review by reitermarkus — DISMISSED** [review]
@dduugg, can you move your last commit to a separate PR? I couldn't really finish reviewing because the “Files changed” tab just hangs.

File: Library/Homebrew/test/spec_helper.rb
Comment (dduugg) [reply Library/Homebrew/test/spec_helper.rb:71]: Also, many potential contributors will not be versed in the Ruby ecosystem, and this should make it clearer what test framework they need to include in their searches. 😄

File: Library/Homebrew/test/spec_helper.rb
Comment (dduugg) [reply Library/Homebrew/test/spec_helper.rb]: Great call, done

**Comment by reitermarkus** [comment]
> let's get this merged before it conflicts.

Merging this quickly is not a good reason to dismiss my request for being able to fully review this. I certainly can't review it any better now that it is merged.

> Remaining feedback can be addressed in follow-up PR

I mean it's generally fine to make further changes to a future PR, but I was not even able to review this fully in order to provide this “remaining feedback”.

**Comment by MikeMcQuaid** [comment]
> Merging this quickly is not a good reason to dismiss my request for being able to fully review this.

You did provide a round of review here and your comments were addressed.

> I certainly can't review it any better now that it is merged.

You can definitely review PRs post-merge, I do this all the time. If there's problems: they can be addressed in a future PR.

> I mean it's generally fine to make further changes to a future PR, but I was not even able to review this fully in order to provide this “remaining feedback”.

Another maintainer (me) was able to review it before it was merged.

**Comment by reitermarkus** [comment]
> You can definitely review PRs post-merge

Sure, not if “Files changed” tab just hangs though, which is why I specifically requested to split this up.

At least the individual commits seem to work now, they were also bugged by not showing the `+` button for commenting last time.

File: Library/Homebrew/test/spec_helper.rb
Comment (reitermarkus) [reply Library/Homebrew/test/spec_helper.rb:71]: I am looking forward to the many potential contributors that will actually write a test. 😄

File: Library/Homebrew/test/rubocops/shell_commands_spec.rb
Comment (reitermarkus) [inline Library/Homebrew/test/rubocops/shell_commands_spec.rb:216]: Can we unnest these instead?

**Pull request merged**
